### ISSUE TITLE
fiona cannot find file in windows path with `!` character

### ISSUE THREAD

**Issue by smnorris** [Toblerity/Fiona#742]
## Expected behavior and actual behavior.
`fiona` cannot find file with `!` in path,
For example, `fiona` seems to interpret `!test\test.geojson` as `test est.geojson`

`click` validates the path successfully in my script with:
`@click.option(... type=click.Path(exists=True))`

Based on https://ss64.com/nt/syntax-esc.html I tried escaping the `!` with `^` and `^^` with no luck, `click` then tells me the path does not exist. Perhaps there is a better way to escape the character?

I can just copy data over to a location with a more reasonable path name but thought this might be worth reporting.

## Steps to reproduce the problem.

```
> mkdir !test
> copy test.geojson !test\test.geojson
> python

Python 3.6.5 |Anaconda, Inc.| (default, Mar 29 2018, 13:32:41) [MSC v.1900 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import fiona
>>> with fiona.open("!test\test.geojson") as src:
...     src.crs
...
Traceback (most recent call last):
  File "fiona\_shim.pyx", line 73, in fiona._shim.gdal_open_vector
  File "fiona\_err.pyx", line 270, in fiona._err.exc_wrap_pointer
fiona._err.CPLE_OpenFailedError: 'test  est.geojson' does not exist in the file system, and is not recognized as a supported dataset name.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "Q:\projects\Mempr\Wind_Mapping\scripts\windsites\venv_36\lib\site-packages\fiona\env.py", line 406, in wrapper
    return f(*args, **kwargs)
  File "Q:\projects\Mempr\Wind_Mapping\scripts\windsites\venv_36\lib\site-packages\fiona\__init__.py", line 257, in open
    layer=layer, enabled_drivers=enabled_drivers, **kwargs)
  File "Q:\projects\Mempr\Wind_Mapping\scripts\windsites\venv_36\lib\site-packages\fiona\collection.py", line 159, in __init__
    self.session.start(self, **kwargs)
  File "fiona\ogrext.pyx", line 484, in fiona.ogrext.Session.start
  File "fiona\_shim.pyx", line 80, in fiona._shim.gdal_open_vector
fiona.errors.DriverError: 'test est.geojson' does not exist in the file system, and is not recognized as a supported dataset name.
>>>
```

## Operating system

Windows Server 2016 Standard
Version 1607 

## Fiona and GDAL version and provenance
Pre-built from https://www.lfd.uci.edu/~gohlke/pythonlibs/ 
`Fiona-1.8.6-cp36-cp36m-win_amd64.whl`
`GDAL-2.3.3-cp36-cp36m-win_amd64.whl`

**Comment by sgillies** [comment]
@smnorris I don't have a windows computer to test this on, but I think you need to use a raw string to represent your filename. Otherwise "\t" is interpreted as a tab.

`with fiona.open(r"!test\test.geojson") as src:`

**Comment by smnorris** [comment]
Ah of course, thanks. I forgot about the raw strings issue.

Unfortunately it still fails - the tab is gone, but the `!` is still not coming through.
I tried all these with the same result:

`with fiona.open(r"!test/test.geojson") as src:`
`with fiona.open(r"^!test/test.geojson") as src:`
`with fiona.open(r"^^!test/test.geojson") as src:`

I'm working around by copying the file to `tempfile.mkdtemp()` via `ogr2ogr`.

```
>>> import fiona
>>> with fiona.open(r"!test/test.geojson") as src:
...     src.crs
...
Traceback (most recent call last):
  File "fiona\_shim.pyx", line 73, in fiona._shim.gdal_open_vector
  File "fiona\_err.pyx", line 270, in fiona._err.exc_wrap_pointer
fiona._err.CPLE_OpenFailedError: test/test.geojson: No such file or directory

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "Q:\projects\Mempr\Wind_Mapping\scripts\windsites\venv_36\lib\site-packages\fiona\env.py", line 406, in wrapper
    return f(*args, **kwargs)
  File "Q:\projects\Mempr\Wind_Mapping\scripts\windsites\venv_36\lib\site-packages\fiona\__init__.py", line 257, in open
    layer=layer, enabled_drivers=enabled_drivers, **kwargs)
  File "Q:\projects\Mempr\Wind_Mapping\scripts\windsites\venv_36\lib\site-packages\fiona\collection.py", line 159, in __init__
    self.session.start(self, **kwargs)
  File "fiona\ogrext.pyx", line 484, in fiona.ogrext.Session.start
  File "fiona\_shim.pyx", line 80, in fiona._shim.gdal_open_vector
fiona.errors.DriverError: test/test.geojson: No such file or directory
```

**Comment by smnorris** [comment]
moving the file to `test\!test\test.geojson`:
```
>>> with fiona.open(r"test\!test\test.geojson") as src:
...     src.crs
...
.
.
.
fiona.errors.DriverError: test\test.geojson: No such file or directory
```

**Comment by sgillies** [comment]
Ah, I found a bug in fiona's path parsing: `!` causes the path to be parsed as though it were a zip archive. The workaround is to avoid `!` in your filenames until we can fix this and make another release.

**Comment by snorfalorpagus** [comment]
Closed by #773.
